pkg/cluster/actions.go

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+
+	"sigs.k8s.io/kind/pkg/cluster/config"
+)
+
+// Action define a set of tasks to be executed on a `kind` cluster.
+// Usage of actions allows to define repetitive, high level abstractions/workflows
+// by composing lower level tasks
+type Action interface {
+	// Tasks returns the list of task that are identified by this action
+	// Please note that the order of task is important, and it will be
+	// respected during execution
+	Tasks() []Task
+}
+
+// Task define a logical step of an action to be executed on a `kind` cluster.
+// At exec time the logical step will then apply to the current cluster
+// topology, and be planned for execution zero, one or many times accordingly.
+type Task struct {
+	// Description of the task
+	Description string
+	// TargetNodes define a function that identifies the nodes where this
+	// task should be executed
+	TargetNodes NodeSelector
+	// Run the func that implements the task action
+	Run func(ec *execContext, configNode *config.Node) error
+}
+
+// NodeSelector defines a function returning a subset of nodes where tasks
+// should be planned.
+type NodeSelector func(*config.Config) config.NodeList
+
+// PlannedTask defines a Task planned for execution on a given node.
+type PlannedTask struct {
+	// task to be executed
+	Task Task
+	// node where the task should be executed
+	Node *config.Node
+
+	// PlannedTask should respects the given order of actions and tasks
+	actionIndex int
+	taskIndex   int
+}
+
+// ExecutionPlan contain an ordered list of Planned Tasks
+// Please note that the planning order is critical for providing a
+// predictable, "kubeadm friendly" and consistent execution order.
+type ExecutionPlan []*PlannedTask
```

Comment:
questions:
- is the node blocking until a planning tasks finishes before executing the next planned task?
i'm yet to see the rest of the code but tasks should be added in a queue of sorts.
- what if a task blocks indefinitely, is there a way to get out of that?

Reply:
ExecutionPlan is an ordered list, and the current execution model is sequential
If something blocks, the way to get out of this is ctrl+c (same as the current version of kind)